### DISCUSSION TITLE
Utilize MQTT integration in Home Assistant directly

### DISCUSSION THREAD

**Discussion by rohankapoorcom** [sca075/mqtt_vacuum_camera#17] [Ideas]
Hi @sca075,

Thanks for putting together this integration. It's super helpful to be able to utilize other vacuum cards rather than the Valetudo specific one only. While going through the setup of this integration, I noticed that it looks for the MQTT User Name / Password and the topic prefix for the vacuum. I wanted to suggest that instead of requesting that from the user (since the user will already be using MQTT in Home Assistant) that it should depend on the MQTT integration in Home Assistant and read/write to MQTT through those functions that already exist. Reference: https://github.com/home-assistant/core/blob/dev/homeassistant/components/mqtt/client.py.

Best!

**Comment by sca075** [comment]
Thank you really much for this, I will have a look at it and will also implement this.. it's a great idea, thank once again 😃

**Reply by rohankapoorcom** [reply]
A related suggestion, you may be able to get the topic information for the vacuum from Home Assistant itself since the MQTT integration must keep track of it and link it to the vacuum entity somewhere. That would make it so the user just provides the entity and then the integration sets up the rest.

**Reply by sca075** [reply]
@rohankapoorcom  would be interesting to do so.. anyhow please allow me to give you the background  of some implementation decisions I made in the past. The connector.py is actually creating a client with paho mqtt that access he topics of the vacuum publish to the common broker.  Initially I thought I could directly connect, as per we are in HA and as you propose here, the broker internally and get the data we need for it (this is still part of the specs). Anyhow after a while an (a lot of failures honestly) I thought.. using paho (this is why we require the credentials and topic) it is also possible to do other thinks.. like a kind of pass troughs between the vacuum and HA.. then.. as per the camera become slowly growing and somehow working I loused focus on this initial part of the development. I think anyhow if we can re-root the data via broker without client section.. this would be a huge improvement and actually as per initially intended to do.  My current goal is to complete the camera functions first, and to be precise I still have two more functions on the list to be done. This anyhow do not mean, that user friendly way... yes I do care a lot about it.. and as soon will be time for it from me or someone else like you for example.. to work in parallel on it.  This is open source in the end.. and I just wish we would use the Valetudo vacuums on HA offering a more user friendly way with this integration that is developed taking in account HA community requirements and dependencies and also the friendly and amazing card I do strongly recommend to use with it. As soon I will have a little of time for that (this project is a kinda hobby for me).. I will seriously have a look to the link you provided. Thank once again :)

**Reply by rohankapoorcom** [reply]
Thanks for sharing the details @sca075! I totally understand the desire to make it work first and then optimize and make it easier to use later on. I was commenting on this because I actually was unable to use the integration as written since it takes the credentials and assumes that it's connecting to a broker running on the local host. In my case, mqtt (and others) are running on a separate docker host over the network so the integration is not able to connect.

I was considering making the modifications to pass along the hostname and then realized I would need to pass along the SSL certificate details as well and realized it would likely be easier to use the MQTT wrapper built into Home Assistant instead. I've taken a look at some of the code, I think the implementation would be straightforward to switch to calling https://github.com/home-assistant/core/blob/dev/homeassistant/components/mqtt/client.py instead of Paho-MQTT. I found several examples in the existing code base that do this:

1. https://github.com/home-assistant/core/blob/7827f9ccaea272c37186fdd6630d6ab848bf07a1/homeassistant/components/manual_mqtt/alarm_control_panel.py#L481
2. https://github.com/home-assistant/core/blob/7827f9ccaea272c37186fdd6630d6ab848bf07a1/homeassistant/components/mqtt_json/device_tracker.py#L83

I'm happy to take a look and try to help out with the implementation, but probably won't have time for a couple of weeks. I totally understand the time constraints with hobby projects! No pressure from my side to make any changes :)

**Reply by sca075** [reply]
@rohankapoorcom I need to say thanks for this, and welcome in Team in this case :) I will study this and we will found a solution for this scenario.. to be honest I didn't think about this possible setup of HA.. absolutely this is a great point and I totally agree (hobby or not) we got to found a solution for it. Some of the docs you provided I already had look at them, and somehow I didn't yet catch how to do it but.. we never give up ;) if I get some proposal on how to go ahead here will post an update on this discussion.

**Comment by phrozen77** [comment]
> In my case, mqtt (and others) are running on a separate docker host over the network so the integration is not able to connect.

Ah, and i wondered why it asked at all and didn't make any mention of the hostname to connect to - also explains why it's currently not working :)

So, i may stop looking for the moment 👍

**Reply by sca075** [reply]
Well.. got it.. we will try to do it with V1.4.0 ;) I think it will take a little of work to do it but it is understood that this become now a priority.

**Comment by sca075** [comment]
Okay here is the game plan:
- I'm going to setup on test HA instance (that run on HA OS PI4) an external MQTT via Docker running on a laptop.
- Then connect the my vacuum to this external MQTT. 
From there will check how to grab the data and test the connection with some code that actually I developed for test purpose. I Think within this weekend will get an idea of the work to do on this part of the implementation (the connector.py actually).

**Reply by phrozen77** [reply]
Sounds good to me, if you need me to test - let me know!

**Comment by sca075** [comment]
Well @phrozen77 if to setup MQTT you followed similar steps like in this: https://www.homeautomationguy.io/blog/docker-tips/configuring-the-mosquitto-mqtt-docker-container-for-use-with-home-assistant
I think we can also do an easy test.. In the end would be necessary just to add the IP or Host name of the instance of running MQTT. In the connector.py at the moment the IP address of the Broker is setup as a constant:
```python
self._broker = "127.0.0.1"
```
If you change that line with the host name and setup the credential and topic of the vacuum the integration might work. Unless you did try already this way of course. The connector.py is in the valetudo folder of the integration in custom_components.
IP address would be preferred but I know for a fact that 127.0.0.1 can be replaced by "localhost" therefore I guess also the hostname should work... it isn't of course a solution but if you are willing to help would be really appreciated if you give it a try.

**Reply by phrozen77** [reply]
That was almost too easy - worked on the first try:

<img width="399" alt="Bildschirmfoto 2023-08-20 um 06 40 38" src="https://github.com/sca075/valetudo_vacuum_camera/assets/1218594/1e80bab0-49f4-4d64-bc03-1736462a068a">

**Reply by phrozen77** [reply]
To add: both tries worked - directly with IP but also hostname as well!

**Reply by sca075** [reply]
Super.. So will add the parameter on the configuration from the next version on.. @rohankapoorcom it looks technically this could be a easy solution any how for the automatic configuration of those parameter.. well I would need a little help and you are very skilled on this.. when you have time would be wonderful if we can easy the setup of the camera with the automatic recognition of those parameter and credentials. Will keep this as a milestone :) thank you to you both guys for raising this point 👍

**Reply by sca075** [reply]
@phrozen77 v1.3.4 release adding the required changes that actually you tested and confirmed thanks a lot for that. 
I hope this will make the life easier not only to us guys :) Thanks a lot once again.

**Reply by rohankapoorcom** [reply]
>  it looks technically this could be a easy solution any how for the automatic configuration of those parameter.. well I would need a little help and you are very skilled on this.. when you have time would be wonderful if we can easy the setup of the camera with the automatic recognition of those parameter and credentials. 

@sca075 I have a decent idea of how to do this, I can poke around maybe later in the week or next weekend. But I'm concerned about breaking things since several unit tests are not passing today. Do you think you can fix those first so that I have more confidence that my changes don't break other things?

**Reply by sca075** [reply]
@rohankapoorcom I'm aboard this week, I will be back in this weekend. I personally check the integration on a dedicated pi4 (the pytest was not updated since a while) I will check how we can manage with this, and confirm it as soon ;)

**Comment by sca075** [comment]
okay the step 1 is now completed.. still on progress of course but in the end just to update you guys the workaround I tested so far is to use instead of localhost the core-mosquito host name (that who use the HA embedded MQTT would not need to change).
![Screenshot 2023-08-20 at 09 01 20](https://github.com/sca075/valetudo_vacuum_camera/assets/82227818/b1350da4-bd38-486e-8411-9de3c02ababb)

I'm going now to change the code and as soon I test all changed parts will push the update on the repository. So in the end when you both agree I would release this modification later today officially on v1.3.4 instead of v1.4.0.

**Reply by phrozen77** [reply]
Sounds good to me!

**Reply by sca075** [reply]
@rohankapoorcom if you will try and it works also for you the only open point would be to make easier the configuration as you proposed, but let's consider that using paho mqtt the client implementation works.. so it just a matter to get MQTT telling us what we area looking for the configuration ;) thanks for confirming if it works also in your case ...

**Reply by rohankapoorcom** [reply]
It works for me as well! Having some trouble figuring out how to get the room info for the card, but I'll open another discussion for that.

**Reply by sca075** [reply]
Tray by settings up a new card.. if you see the attributes listed with rooms.. then with a new card generate room configuration will pop out the rooms when you select rooms cleaning.. then outline isn't yet 100% okay you might need to adjust some value.. hope this helps:)

**Reply by rohankapoorcom** [reply]
No luck unfortunately. The generate room configuration didn't pop up and the button to trigger it is not click able.

**Reply by sca075** [reply]
did you confirm if in the status attributes of the camera there is rooms and the list of the rooms you configured? 
What is your Robot type? Can you please provide me the yourvacuum.json in the screenshot at gsca075[at gmail.com]. 
by the way the data are collected by "segments" and tested it with the Dreame D9 json.. image_handler has a function that create the values that is later on formatted in the state_attributes of the camera.yourvacuum.

**Reply by rohankapoorcom** [reply]
Email sent with the details. Thanks for taking a look!

**Reply by sca075** [reply]
@rohankapoorcom no problem, I'm not sure if you got the email I sent you back. Config/custom_components/valetudo_vacuum_camera/snapshots is where you find the your_vacuum.json that is actually the decoded payload the integration use to produce the image. 

Note, that the card cannot create the rooms configuration because there's no camera rooms attribute is output (it is so programmed when the data are not available) I would need to check why the rooms aren't recognized, probably some keyword difference between the one hardcoded in the image_handler function that generates the data. 

This data's I'm here requesting are only for debugging this function.

Thank you for your understanding and cooperation here.
